File: Assets/Scripts/FPSCounter.js
#pragma strict

/* dependencies */
var fpsCounter : UI.Text;

/* globals */
var updateInterval = 0.5;

/* scratchpad */
private var showFPS : boolean = false;

private var accum = 0.0; // FPS accumulated over the interval
private var frames = 0; // Frames drawn over the interval
private var timeleft : float; // Left time for current interval

private var badFrames : Hashtable = new Hashtable(); // keyed off timeCounter

@script RequireComponent(UI.Text)
function Start () {
	fpsCounter = GetComponent(UI.Text);
	timeleft = updateInterval;
}

function Update () {
	if (Input.GetKeyDown(KeyCode.F)) {
		showFPS = !showFPS;
	}

	timeleft -= Time.deltaTime;
    accum += Time.timeScale/Time.deltaTime;
    ++frames;
 
    // Interval ended - update GUI text and start new interval
    if (timeleft <= 0.0) {
        var fps : float = (accum/frames);
        var tc : int = Mathf.FloorToInt(Time.time);
        
        if (fps < 57) {
        	if (badFrames.ContainsKey(tc)) {
        		var x : Array = badFrames[tc];
        		x.Push(fps.ToString("f1"));
        		badFrames[tc] = x;
        	} else {
        		x = Array();
        		x[0] = fps.ToString("f1");
        		badFrames[tc] = x;
        	}
        }
		if (tc != 0 && tc % 30 == 0 && MetricsLevel.level.CompareTo(MetricsLevel.VERBOSE) == 0) {
			EnemySpawnManager.gaLogEvent("Application", "BadFrames@" + tc + ":" + getBadFrames());
		}
        
        if (fpsCounter && showFPS) fpsCounter.text = fps.ToString("f1") + " fps";
        timeleft = updateInterval;
        accum = 0.0;
        frames = 0;
    }
    
    if (!showFPS) {
    	fpsCounter.text = "";
    }
}

private function getBadFrames() {
	var fData = "{";
	for (var i = 0; i < 600; i++) {
		if (badFrames.ContainsKey(i)) {
			fData += i + ': "' + badFrames[i] + '",';
		}
	}
	fData += "}";
	return fData;
}

File: Assets/Scripts/messages/FPSCounter.cs
using UnityEngine;
using System.Collections;

[System.Serializable]
/**
 * @author: cmdr2 <secondary.cmdr2@gmail.com>
 */
/* dependencies */
/* globals */
/* scratchpad */
 // FPS accumulated over the interval
 // Frames drawn over the interval
 // Left time for current interval
 // keyed off timeCounter
[UnityEngine.RequireComponent(typeof(UnityEngine.UI.Text))]
public partial class FPSCounter : MonoBehaviour
{
    public UnityEngine.UI.Text fpsCounter;
    private static float updateInterval;
    private bool showFPS;
    private float accum;
    private int frames;
    private float timeleft;
    //private Hashtable badFrames;
    public virtual void Start()
    {
        this.fpsCounter = (UnityEngine.UI.Text) this.GetComponent(typeof(UnityEngine.UI.Text));
        this.timeleft = FPSCounter.updateInterval;
    }

    public virtual void Update()
    {
        if (Input.GetKeyDown(KeyCode.F))
        {
            this.showFPS = !this.showFPS;
        }
        this.timeleft = this.timeleft - Time.deltaTime;
        this.accum = this.accum + (Time.timeScale / Time.deltaTime);
        ++this.frames;
        // Interval ended - update GUI text and start new interval
        if (this.timeleft <= 0f)
        {
            float fps = this.accum / this.frames;
            int tc = Mathf.FloorToInt(Time.time);
            /*if (fps < 57)
            {
                if (this.badFrames.ContainsKey(tc))
                {
                    object[] x = (object[]) this.badFrames[tc];
                    x.Push(fps.ToString("f1"));
                    this.badFrames[tc] = x;
                }
                else
                {
                    x = new object[0];
                    x[0] = fps.ToString("f1");
                    this.badFrames[tc] = x;
                }
            }
            if (((tc != 0) && ((tc % 30) == 0)) && (MetricsLevel.level.CompareTo(MetricsLevel.VERBOSE) == 0))
            {
                Metrics.LogEvent("Application", (("BadFrames@" + tc) + ":") + this.getBadFrames());
            }*/
            if (this.fpsCounter && this.showFPS)
            {
                this.fpsCounter.text = fps.ToString("f1") + " fps";
            }
            this.timeleft = FPSCounter.updateInterval;
            this.accum = 0f;
            this.frames = 0;
        }
        if (!this.showFPS)
        {
            this.fpsCounter.text = "";
        }
    }

    /*private string getBadFrames()
    {
        string fData = "{";
        int i = 0;
        while (i < 600)
        {
            if (this.badFrames.ContainsKey(i))
            {
                fData = fData + (((i + ": \"") + this.badFrames[i]) + "\",");
            }
            i++;
        }
        fData = fData + "}";
        return fData;
    }*/

    public FPSCounter()
    {
        //this.badFrames = new Hashtable();
    }

    static FPSCounter()
    {
        FPSCounter.updateInterval = 0.5f;
    }

}